File: backend/app/models/google_location.py
from sqlalchemy import Column, String, DateTime, Text, ForeignKey, Integer, Boolean, Enum as SQLEnum
from sqlalchemy.sql import func
from app.core.database import Base
import uuid
import enum


class LocationStatus(enum.Enum):
    """Google Business location status"""
    VERIFIED = "Verified"
    NEEDS_ATTENTION = "Needs attention"
    LIMITED_ACCESS = "Limited access"
    NOT_ELIGIBLE = "Not eligible"


class GoogleLocation(Base):
    """Google Business Profile location"""
    __tablename__ = "google_locations"
    
    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    business_profile_id = Column(String(36), ForeignKey("business_profiles.id", ondelete="SET NULL"), index=True)
    
    # Google Business Profile data
    google_location_id = Column(String(255), unique=True, nullable=False, index=True)
    name = Column(String(255), nullable=False)
    category = Column(String(255))
    primary_location = Column(String(255))
    website = Column(String(500))
    
    # Location details
    city = Column(String(100))
    state = Column(String(100))
    country = Column(String(100))
    full_address = Column(Text)
    
    # Status and metrics
    status = Column(SQLEnum(LocationStatus), nullable=False)
    review_count = Column(Integer, default=0)
    average_rating = Column(String(10))
    
    # Sync status
    is_synced = Column(Boolean, default=False, nullable=False)
    
    # Timestamps
    last_fetched_at = Column(DateTime(timezone=True))
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    
    def __repr__(self):
        return f"<GoogleLocation(id={self.id}, name={self.name}, status={self.status})>"

File: backend/src/models/GoogleLocation.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface IGoogleLocation extends Document {
  _id: Types.ObjectId;
  userId: Types.ObjectId;
  businessProfileId?: Types.ObjectId;
  googleLocationId: string;
  name: string;
  category?: string;
  primaryLocation?: string;
  website?: string;
  city?: string;
  state?: string;
  country?: string;
  fullAddress?: string;
  status: string;
  reviewCount: number;
  averageRating?: string;
  isSynced: boolean;
  lastFetchedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const googleLocationSchema = new Schema<IGoogleLocation>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    businessProfileId: {
      type: Schema.Types.ObjectId,
      ref: 'BusinessProfile',
      index: true,
    },
    googleLocationId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    name: {
      type: String,
      required: true,
    },
    category: String,
    primaryLocation: String,
    website: String,
    city: String,
    state: String,
    country: String,
    fullAddress: String,
    status: {
      type: String,
      enum: ['VERIFIED', 'NEEDS_ATTENTION', 'LIMITED_ACCESS', 'NOT_ELIGIBLE'],
      default: 'VERIFIED',
    },
    reviewCount: {
      type: Number,
      default: 0,
    },
    averageRating: String,
    isSynced: {
      type: Boolean,
      default: false,
    },
    lastFetchedAt: Date,
  },
  {
    timestamps: true,
  }
);

export const GoogleLocation = model<IGoogleLocation>(
  'GoogleLocation',
  googleLocationSchema
);
